pkg/azuredx/datasource.go

```diff
@@ -1,30 +1,101 @@
-package main
+package azuredx
 
 import (
 	"fmt"
 
-	"github.com/grafana/azure-data-explorer-datasource/pkg/azuredx"
+	"github.com/grafana/azure-data-explorer-datasource/pkg/azuredx/client"
+	"github.com/grafana/azure-data-explorer-datasource/pkg/azuredx/models"
+	"github.com/grafana/azure-data-explorer-datasource/pkg/azuredx/tokenprovider"
 	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	sdkhttpclient "github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/instancemgmt"
 	"github.com/grafana/grafana-plugin-sdk-go/data"
 	jsoniter "github.com/json-iterator/go"
 	"golang.org/x/net/context"
 )
 
 // GrafanaAzureDXDatasource stores reference to plugin and logger
-type GrafanaAzureDXDatasource struct{}
+type GrafanaAzureDXDatasource struct {
+	client   *client.Client
+	settings *models.DatasourceSettings
+}
+
+func NewDatasource(settings backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
+	datasourceSettings := models.DatasourceSettings{}
+	err := datasourceSettings.Load(settings)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenProvider := tokenprovider.NewAccessTokenProvider(datasourceSettings.ClientID, datasourceSettings.TenantID, "AzurePublic", datasourceSettings.Secret, []string{"https://kusto.kusto.windows.net/.default"})
+
+	httpClientProvider := sdkhttpclient.NewProvider(sdkhttpclient.ProviderOptions{
+		Middlewares: []sdkhttpclient.Middleware{
+			client.AuthMiddleware(tokenProvider),
+		},
+	})
+
+	httpClientOptions, err := settings.HTTPClientOptions()
+	if err != nil {
+		backend.Logger.Error("failed to create HTTP client options", "error", err.Error())
+	}
+
+	httpClient, err := httpClientProvider.New(httpClientOptions)
+	if err != nil {
+		backend.Logger.Error("failed to create HTTP client", "error", err.Error())
+	}
+
+	return &GrafanaAzureDXDatasource{
+		client:   client.New(httpClient),
+		settings: &datasourceSettings,
+	}, nil
```

Comment:
Not totally familiar with Go and it's practices, but should this return an error if it can't make an http client?

Reply:
yes!!